pkg/server/bundle/client/updater.go

```diff
@@ -0,0 +1,122 @@
+package client
+
+import (
+	"context"
+	"crypto/x509"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/pkg/common/bundleutil"
+	"github.com/spiffe/spire/pkg/common/idutil"
+	"github.com/spiffe/spire/pkg/common/pemutil"
+	"github.com/spiffe/spire/proto/spire/server/datastore"
+	"github.com/zeebo/errs"
+)
+
+type BundleUpdaterConfig struct {
+	TrustDomainConfig
+
+	TrustDomain string
+	Log         logrus.FieldLogger
+	DataStore   datastore.DataStore
+
+	// newClient is a test hook for injecting client behavior
+	newClient func(ClientConfig) Client
+}
+
+type BundleUpdater interface {
+	UpdateBundle(ctx context.Context) (time.Duration, error)
+}
+
+type bundleUpdater struct {
+	c      BundleUpdaterConfig
+	client Client
+	bundle *bundleutil.Bundle
+}
+
+func NewBundleUpdater(ctx context.Context, config BundleUpdaterConfig) (BundleUpdater, error) {
+	if config.newClient == nil {
+		config.newClient = func(config ClientConfig) Client {
+			return NewClient(config)
+		}
+	}
+
+	// Load if the existing bundle, if present, and initialize the client.
+	var rootCAs []*x509.Certificate
+	bundle, err := fetchBundle(ctx, config.DataStore, config.TrustDomain)
+	switch {
+	case err != nil:
+		return nil, errs.Wrap(err)
+	case bundle != nil:
+		// There is an existing bundle for the trust domain.
+		rootCAs = bundle.RootCAs()
+	case bundle == nil:
+		// No current bundle for the trust domain. Load roots from the bootstrap bundle.
+		rootCAs, err = pemutil.LoadCertificates(config.BootstrapBundle)
+		if err != nil {
+			return nil, errs.New("unable to load bootstrap bundle: %v", err)
+		}
+	}
+
+	u := &bundleUpdater{
+		c:      config,
+		bundle: bundle,
+	}
+	u.client = u.newClient(rootCAs)
+	return u, nil
+}
+
+func (u *bundleUpdater) UpdateBundle(ctx context.Context) (time.Duration, error) {
+	u.c.Log.Debug("Polling for federated bundle updates")
+	bundle, metadata, err := u.client.FetchBundle(ctx)
+	if err != nil {
+		u.c.Log.WithError(err).Error("Failed to poll for federated bundle update")
+		return 0, err
+	}
+
+	switch {
+	case u.bundle == nil:
+		u.c.Log.Info("Creating federated bundle")
+		_, err = u.c.DataStore.CreateBundle(ctx, &datastore.CreateBundleRequest{
+			Bundle: bundle.Proto(),
+		})
+	case !u.bundle.EqualTo(bundle):
+		u.c.Log.Info("Updating federated bundle")
+		_, err = u.c.DataStore.UpdateBundle(ctx, &datastore.UpdateBundleRequest{
+			Bundle: bundle.Proto(),
+		})
+	default:
+		return metadata.RefreshHint, nil
+	}
+	if err != nil {
+		u.c.Log.WithError(err).Error("Failed to store federated bundle update")
+		return 0, errs.Wrap(err)
+	}
+
+	u.client = u.newClient(bundle.RootCAs())
```

Comment:
I think some of the code here and in the client may be problematic if there is more than one spire server configured for federation (which IMO should be supported), mostly due to us holding the bundles in memory.

Will DataStore.UpdateBundle fail if the td doesn't already have a bundle defined? I could see a couple servers racing each other on boot.

Maybe racing for updates doesn't matter so much? If two servers race for an update, one gets an "old" copy (e.g. without a new key published) and one gets a "new" copy, and the "old" copy wins the write... then it could be assumed that we would pick up the new key at maximum spiffe_refresh_hint?

One other thing we could do is require a bundle to be provided with the call to client.FetchBundle. Then this function could get the latest bundle from the datastore, use it to pull a new copy, and write it back if it doesn't match the one we just fetched.

I'm sure there are other ways to handle this, just thinking out loud on all the cases I can conjur up here

Reply:
Yeah, it is racy. I'm not sure how we'd solve it without implementing versioning in the datastore. Even with the method you described there is nothing preventing another server from updating the bundle between the current server's "fetch, poll, write if it hasn't changed" cycle.